### PR TITLE
#164691220 Add User profiles Functionality              

### PR DESCRIPTION
**Pull request by NicholusMuwonge**
#### What does this PR do?
Enables user to create a profile and also update it.

#### Description of Task to be completed?
*To get the following end points working*:
- ```**PUT** http://127.0.0.1:8000/api/user/```
- ```**GET** http://127.0.0.1:8000/api/user/``` 
- ```**GET** http://127.0.0.1:8000/api/profiles/<str:username>/```
- ```**GET** http://127.0.0.1:8000/api/profiles/```

#### How should this be manually tested?
**One** should be able to clone this repo and test out the code In Postman or any other tool of their choice to test it out.

- Added the relational model which creates a profile on the creation of a user.
- The user can update their profile if they have the right authorization using the route ```http://127.0.0.1:8000/api/user/ ``` 
The updated body should look like.
```
{
  "user":{
    "username": "nicksbros",
    "email": "some@some.com",
    "bio":"I work at a statefarm i",
    "image": "https://inspector"
    
}
}
```
- User can access their own profile using a get method using routes ```http://127.0.0.1:8000/api/user/``` or ``` http://127.0.0.1:8000/api/profiles/<str:username> ```
response should look like 
```
{
    "user": {
        "email": "some@some.com",
        "username": "nicksbros",
        "token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpZCI6MSwidXNuIjoibmlja3Nicm9zIiwiZXhwIjoxNTU2OTk2MTkyfQ.W4XnFcVppUCfj6_vhF6TuGjXxoYCzbNDe1EeN6mP4E0",
        "bio": "I work at a statefarm i",
        "image": ""
    }
}
```
the image field shoul be black since it only returns an image object.

- The user should be able to view other people's profiles using route ``` http://127.0.0.1:8000/api/profiles/<username> ``` and the response object should look like 

``` {
    "profile": {
        "username": "nicksbrosty",
        "bio": "I am cool bro",
        "image": "https://static.productionready.io/images/smiley-cyrus.jpg"
    }
}
```
with a default route which represents the avatar to act as a place holder for the user's profile picture. 
- The user shouldn't be able to edit one's profile as for the PUT method for this route is blocked. like
```{
    "profile": {
        "detail": "Method \"PUT\" not allowed."
    }
}
```
- If user doesnt exist, response should look like ``` {
    "errors": {
        "detail": "The requested profile does not exist."
    }
}```

- A time stamp for created_at and updated_at

**Story Id** https://www.pivotaltracker.com/story/show/164691220

### CODE REVIEW

**Review by RonKbS — CHANGES_REQUESTED** [review]
Please share your reason for having commented out code such as on lines 80 and 83 in the models file of the authentication app. The same for line 7 in renderers.py of the same app.

**Comment by NicholusMuwonge** [comment]
> Please share your reason for having commented out code such as on lines 80 and 83 in the models file of the authentication app. The same for line 7 in renderers.py of the same app.

for both of these since models and profiles share created-at and updated at, I created a model ordering them in a way created at comes 1st and the other next to make more sense and to make the user authentication table in sync with the user_profile table. so I created a profile in core/models.py from which both of these inherit so not commenting those lines would make no sense coz they're already taken care of from the core where profile model and user models inherit from.

This also goes for renderers.py the content would be needed for any other apps created so duplicating isn't the best practice so I placed it in the cores where both profiles app and authentication app can inherit from and if you observe both renderers profiles and authentication they import from the core renderers.

I hope this isnt too much wording to explain your question please. Thanks

File: authors/apps/authentication/renderers.py [hunk 1]
```diff
@@ -1,24 +1,14 @@
 import json
 
-from rest_framework.renderers import JSONRenderer
+# from rest_framework.renderers import JSONRenderer
+from authors.apps.core.renderers import AuthorsJSONRenderer
```
Comment (reifred) [inline authors/apps/authentication/renderers.py:3]: Remove this commented code.

File: authors/apps/authentication/renderers.py [hunk 2]
```diff
@@ -1,24 +1,14 @@
...
-            # As mentioned about, we will let the default JSONRenderer handle
-            # rendering errors.
-            return super(UserJSONRenderer, self).render(data)
-
-        # Finally, we can render our data under the "user" namespace.
+        
         return json.dumps({
             'user': data
         })
+
```
Comment (reifred) [inline authors/apps/authentication/renderers.py:9]: Since you are abstracting AuthorsJSONRender this function is not necessary just stop on overriding the object_label everything thing will be catered for.

File: authors/apps/authentication/views.py [hunk 3]
```diff
@@ -89,7 +89,17 @@ def retrieve(self, request, *args, **kwargs):
 
     def update(self, request, *args, **kwargs):
         serializer_data = request.data.get('user', {})
-
+        user_data = request.data.get('user', {})
+
```
Comment (reifred) [inline authors/apps/authentication/views.py]: You are having two variables doing the same thing. I think one should be removed.

File: authors/apps/profiles/serializers.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    image = serializers.CharField(allow_blank=True, required=False)
+
+    class Meta:
+        model = Profile
+        fields = ('username', 'bio', 'image',)
+        read_only_fields = ('username',)
+
+    def get_image(self, obj):
+        if obj.str:
+            return obj.str
```
Comment (reifred) [inline authors/apps/profiles/serializers.py]: You are not using this function anywhere since you commented out the image field it used as a serializer method. Also, the image field is allowing any kinds of strings since it's declared as **Charfield** in your serializers yet you specified it to be URLField in your model.
Example;
```
This shouldn't be right as specified in your PR because of the invalid link in your image data.
{
  "user":{
    "username": "nicksbros",
    "email": "some@some.com",
    "bio":"I work at a statefarm i",
    "image": "https://inspector"
    
}
But this is right.
{
  "user":{
    "username": "nicksbros",
    "email": "some@some.com",
    "bio":"I work at a statefarm i",
    "image": "https://inspector.jpg"
    
}
```
I think making use of the serializer method you commented out will solve this problem but you don't include the brackets since it is not taking in any method_name.
So this would be right.
```
image = serializers.SerializerMethodField
```

**Review by reifred — CHANGES_REQUESTED** [review]
@NicholusMuwonge thanks for the work done. The feature was working fine as specified though I found out some issues which I thought would be addressed. Consider including a blank line at the end of your files and other issues are specified below.

**Comment by NicholusMuwonge** [comment]
Thanks @reifred, let me get started working on those issues asap

File: authors/apps/profiles/serializers.py (hunk 4 above)
Comment (NicholusMuwonge) [reply authors/apps/profiles/serializers.py]: This will work in a way that It won't allow any input from the user if it's not an image object so, for now, this will keep rendering and empty image field regardless of if someone included it in the request body or not. Which case I cant test in postman at this point

File: authors/apps/authentication/views.py (hunk 3 above)
Comment (NicholusMuwonge) [reply authors/apps/authentication/views.py]: Sorry, I had to change it, I just forgot to take one out but I have fixed it thanks for the review @reifred

File: authors/apps/authentication/renderers.py (hunk 2 above)
Comment (NicholusMuwonge) [reply authors/apps/authentication/renderers.py:9]: I have worked on this . Thanks

File: authors/apps/authentication/renderers.py (hunk 1 above)
Comment (NicholusMuwonge) [reply authors/apps/authentication/renderers.py:3]: I have , please check back and see the changes

File: authors/apps/authentication/__init__.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+class AuthenticationAppConfig(AppConfig):
+    name = 'authors.apps.authentication'
+    label = 'authentication'
+    verbose_name = 'Authentication'
+
+    def ready(self):
+        import authors.apps.authentication.signals
+
+# This is how we register our custom app config with Django. Django is smart
```
Comment (RonKbS) [inline authors/apps/authentication/__init__.py]: When I read these comments, I get the impression that you've copied and pasted this code.
Please ensure that when you're using code from a tutorial, you understand what it does and write it accordingly.
In this case that means you write (in layman's terms) what the introduction of this code into this `__init__.py` file does.

I also see missing new lines in this file and one or two others.
This would affect you negatively if you were working on a real product because it is not quality code.
Kindly make the necessary changes.

**Review by RonKbS — CHANGES_REQUESTED** [review]

File: authors/apps/authentication/__init__.py (hunk 5 above)
Comment (NicholusMuwonge) [reply authors/apps/authentication/__init__.py]: The code in there helps for automation of Django to automatically pick up a signal every time a new user is created and hence use it to create an associated profile. Let me make more research on the would be missing lines please.

Comment (RonKbS) [reply authors/apps/authentication/__init__.py]: There is no need to research about the missing new_lines. You simply have to press enter at the end of this file to get rid of this red mark 👇 
Screenshot 2019-04-10 at 17 54 06

Regarding the comments, please replace it with what you've said above. Reading it, a person can tell that you are the one that wrote the comments, as opposed to what is currently there, from which I garnered that you'd gotten this section of code from somewhere else.

Comment (NicholusMuwonge) [reply authors/apps/authentication/__init__.py]: I have resolved this please check back

File: authors/apps/profiles/urls.py [hunk 6]
```diff
@@ -0,0 +1,8 @@
+from django.urls import path
+
+from .views import ProfileRetrieveAPIView
+
+app_name = 'profiles'
+urlpatterns = [
+    path('profiles/<str:username>', ProfileRetrieveAPIView.as_view()),
```
Comment (RonKbS) [inline authors/apps/profiles/urls.py]: Please add a `/` at the end of this URL and make it protected as well. Follow the implementation that was used in creating articles.
If you're able to before 10 a. m tomorrow, add a protected URL to retrieve all profiles as well.

**Review by RonKbS — CHANGES_REQUESTED** [review]

File: authors/apps/profiles/urls.py (hunk 6 above)
Comment (NicholusMuwonge) [reply authors/apps/profiles/urls.py]: okay on it

Comment (NicholusMuwonge) [reply authors/apps/profiles/urls.py]: I have added this URL please check it out

File: authors/apps/profiles/views.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+from rest_framework.permissions import AllowAny, IsAuthenticated
+from rest_framework.response import Response
+from .exceptions import ProfileDoesNotExist
+from .models import Profile
+from .renderers import ProfileJSONRenderer
+from .serializers import ProfileSerializer, RetrieveProfiles
+
+
+class ProfileRetrieveAPIView(RetrieveAPIView):
+    permission_classes = (AllowAny,)
```
Comment (RonKbS) [inline authors/apps/profiles/views.py]: Please change this permision class to protect the route

File: authors/apps/profiles/urls.py [hunk 8]
```diff
@@ -0,0 +1,10 @@
+from django.urls import path
+
+from .views import ProfileRetrieveAPIView, RetrieveProfilesView
+
+
+app_name = 'profiles'
+urlpatterns = [
+    path('profiles/<str:username>', ProfileRetrieveAPIView.as_view()),
```
Comment (RonKbS) [inline authors/apps/profiles/urls.py]: You have not yet added a slash to this URL on line 8.
Since we're following a convention where all our URLs have / at the end, do the same for this as well.

**Review by RonKbS — CHANGES_REQUESTED** [review]

File: authors/apps/profiles/urls.py (hunk 8 above)
Comment (NicholusMuwonge) [reply authors/apps/profiles/urls.py]: sorry about that, I have fixed it

File: authors/apps/profiles/views.py (hunk 7 above)
Comment (NicholusMuwonge) [reply authors/apps/profiles/views.py]: I have resolved this

**Review by richien — CHANGES_REQUESTED** [review]
Hi @NicholusMuwonge , I tested this locally and it was working fine. The only issues I have found is that there are slight deviations from the [Coding Conventions](https://github.com/aqua-organisation/ah-django-Aqua/wiki/Coding-Conventions-for-ah-django-Aqua#pr-etiquette) in your PR. Kindly refrain from deviating from the format stated in the conventions because it makes our work as a team appear un-uniform. Please stick to the conventions as below:
#### What does this PR do?
#### Description of Task to be completed?
#### How should this be manually tested?
#### Any background context you want to provide?
#### What are the relevant pivotal tracker stories?
#### Screenshots (if appropriate)
#### Questions:
Kindly also re-read through your PR and fix some spelling errors for instance, there is a line where I assume you meant to write `avatars` but wrote `vators` instead. Kindly proofread your PR and fix those small errors. 
Other than that, this is good and I will approve once you have implemented those changes.
NOTE: You can use `Grammarly` in your PR to highlight some of these errors when you are writing.

**Review by RonKbS — APPROVED** [review]

File: authors/apps/profiles/views.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+        serializer = self.serializer_class(profile)
+
+        return Response(serializer.data, status=status.HTTP_200_OK)
+
+
+class RetrieveProfilesView(RetrieveAPIView):
+    permission_classes = (IsAuthenticated,)
+    renderer_classes = (ProfileJSONRenderer,)
+    serializer_class = ProfileSerializer
+
```
Comment (reifred) [inline authors/apps/profiles/views.py:35]: I have found no use of the above serializer given in the same view you are using the serializer below.
```
def retrieve(self,request):
        profiles = Profile.objects.filter().order_by('-created_at')
        serializer = RetrieveProfiles(profiles, many=True)  <-----
        return Response(
            serializer.data,
            status=status.HTTP_200_OK
            )
```
serializers.py
```
class RetrieveProfiles(serializers.ModelSerializer):
    username = serializers.CharField(source='user.username')

    class Meta:
        model = Profile
        exclude =('id','user',)
        read_only_fields = ('username',)
```

File: authors/apps/profiles/serializers.py [hunk 10]
```diff
@@ -0,0 +1,31 @@
+from rest_framework import serializers
+
+from .models import Profile
+
+
+class ProfileSerializer(serializers.ModelSerializer):
+    username = serializers.CharField(source='user.username')
+    bio = serializers.CharField(allow_blank=True, required=False)
+    image = serializers.SerializerMethodField()
+
```
Comment (reifred) [inline authors/apps/profiles/serializers.py:10]: When updating the user profile data. The image is not updated unless you change the above serializer method to `image = serializers.SerializerMethodField` in other words brackets should be removed. Kindly try testing with **http://127.0.0.1:8000/api/user/**   when updating an image to see what I mean.

**Review by reifred — CHANGES_REQUESTED** [review]
@NicholusMuwonge thanks for the work done with the profile feature and working on requested changes. The application is working fine though I found just little issues which I addressed below. Otherwise thanks for the work done.

**Review by richien — CHANGES_REQUESTED** [review]
Hi @NicholusMuwonge, I have tested all the endpoints locally and they work with the exception of updating an image's URL, which I was able to update only after implementing as @reifred  suggested above. Kindly implement the change he suggested. Other than that, thank you for the work done.

**Comment by NicholusMuwonge** [comment]
> Hi @NicholusMuwonge, I have tested all the endpoints locally and they work with the exception of updating an image's URL, which I was able to update only after implementing as @reifred suggested above. Kindly implement the change he suggested. Other than that, thank you for the work done.

I have updated my code please review the changes

**Review by richien — APPROVED** [review]

**Review by reifred — APPROVED** [review]
